### ISSUE TITLE
RED gives validation error for variables file

### ISSUE THREAD

**Issue by mmesnjak** [nokia/RED#63]
Hi,

I constantly get this error:
```
An internal error occurred during: 'Revalidating model'.
Unable to communicate with XML-RPC server. File /home/misel/projects/msbc-validation-tests/vars/capacity_group_vars.py with arguments []
```

when editing a robot file:

```
*** Settings ***
Library    ProvisioningLibrary
Variables    ../vars/capacity_group_vars.py
*** Keywords ***
Carrier Capacity Groups provisioning is requested
    [Arguments]    ${ccg_list}
    Provision Entities    CCG    ${ccg_list}
```

that loads variables from the *.py file:
```
import vn_provisioning_vars as _vnv
import service_provisioning_vars as _sv
import satellite_band_provisioning_vars as _sbv
import provisioning_vars as _pv
import common_vars as _cv
from zeep.transports import Transport as _Transport
from requests.auth import HTTPBasicAuth as _HTTPBasicAuth
from zeep import Client as _Client

_transport = _Transport(http_auth=_HTTPBasicAuth(_cv.USERNAME, _cv.PASSWORD))
_provisioning_interface = _Client(wsdl=_cv.SLM_HUB_PROVISIONING_WSDL, transport=_transport)

_ss_type = _provisioning_interface.get_type('ns0:supportedService')
_vsat_settings_type = _provisioning_interface.get_type('ns0:vSatSettings')
_vsat_correlation_settings_type = _provisioning_interface.get_type('ns0:vSatCorrelationSettings')

_ccg_type = _provisioning_interface.get_type('ns0:carrierCapacityGroup')
_macg_type = _provisioning_interface.get_type('ns0:multipleAccessCapacityGroup')

LIST__CCG_DATA = [
    _ccg_type(entityId='G13-12K', 
              lowerFreqHz=14196000000, 
              upperFreqHz=14204000000, 
              satelliteBandId=_sbv.satBandName, 
              supportedServices={'service': [_ss_type(vn=_vnv.testVnName, 
                                                      serviceClass='Satellite Internet Access - 3100 Return')]},
              vsatSettings=_vsat_settings_type(returnTechnology='HRC_SCPC', 
                                               returnCapacityGroupName='HRC-SCPC-1',
                                               returnCapacityGroupDomainName='hno', 
                                               dynamicCapacityManagement=False),
              vsatCorrelationSettings=_vsat_correlation_settings_type(applicableReturnTechnology='HRC_MXDMA',
                                                                      returnPoolName='PSSI-HRC-Rtn-TB-Pool',
                                                                      returnPoolDomainName='hno')
    ),
]
```

Is there a way around this?

**Comment by KrzysztofJozefowicz** [comment]
Hi
Hard to say what is the root cause as there might be 2 possible options.
First capacity_group_vars.py is using external clients to get something back ( _provisioning_interface ) and than there is a layer of serialization of returned data to RED so either there is something wrong on interface (timouts etc.) or on serialization of LIST__CCG_DATA .  
To rule _provisioning_interface, can you change var file is such a way that static data is used (valid data but avoid calling _provisioning_interface) and try to reproduce it? 
Also can you provide .log file from workspace/.metadata  folder?

**Comment by mmesnjak** [comment]
Unfortunately, I don't know how to mock this SOAP object that I need to construct using SOAP client.
But, I can verify that after the first time that the error message appears on such a variables file. Whatever I do, even if I comment out everything, does not make any difference to the Eclipse.
Also, I can say that if I have normal list object with dictionaries in it this error does not show up.

[RED_workspace.log.txt](https://github.com/nokia/RED/files/634057/RED_workspace.log.txt)

I can attach the whole project export if you want.

**Comment by adam-wypych** [comment]
You do not need I think since logs say what is the root cause:
```
Caused by: org.apache.xmlrpc.XmlRpcException: <type 'exceptions.TypeError'>:cannot marshal <class 'collections.OrderedDict'> objects
	at org.apache.xmlrpc.client.XmlRpcStreamTransport.readResponse(XmlRpcStreamTransport.java:197)
	at org.apache.xmlrpc.client.XmlRpcStreamTransport.sendRequest(XmlRpcStreamTransport.java:156)
	at org.apache.xmlrpc.client.XmlRpcHttpTransport.sendRequest(XmlRpcHttpTransport.java:143)
	at org.apache.xmlrpc.client.XmlRpcSunHttpTransport.sendRequest(XmlRpcSunHttpTransport.java:69)
	at org.apache.xmlrpc.client.XmlRpcClientWorker.execute(XmlRpcClientWorker.java:56)
	at org.apache.xmlrpc.client.XmlRpcClient.execute(XmlRpcClient.java:167)
	at org.apache.xmlrpc.client.XmlRpcClient.execute(XmlRpcClient.java:137)
	at org.apache.xmlrpc.client.XmlRpcClient.execute(XmlRpcClient.java:126)
	at org.rf.ide.core.executor.RobotCommandRcpExecutor.callRpcFunction(RobotCommandRcpExecutor.java:372)
	at org.rf.ide.core.executor.RobotCommandRcpExecutor.getVariables(RobotCommandRcpExecutor.java:247)
```

**Comment by mmesnjak** [comment]
I think this is from before when I used OrderedDict object in one of the variables python files.
Now the problem seems to be with the SOAP objects created by the zeep SOAP client `get_type` methods.

If you want I can delete the .log file and provoke an error and then send you a fresh log file?

**Comment by adam-wypych** [comment]
Yes please delete `.log` file and provoke an error to appear.

**Comment by mmesnjak** [comment]
Well, again it says that the problem is with OrderedDict.
Please see fresh log file in the attachment.
Since I'm not using it in my own python files it must be that zeep SOAP client is using it internally:
```
(r2)misel@misel-amphinicy[~/projects/msbc-validation-tests]:
[10:53:10] (master)* $ grep OrderedDict * -r
(r2)misel@misel-amphinicy[~/projects/msbc-validation-tests]:
[10:53:13] (master)* $
```
[RED_workspace.log.txt](https://github.com/nokia/RED/files/636249/RED_workspace.log.txt)

**Comment by adam-wypych** [comment]
Ok, I see that again is the same problem type.

Could You say:
1) which RED version do You use?
2) what interpreter version You have?
3) which Robot Framework version?

Thanks in advance.

**Comment by mmesnjak** [comment]
Sure, I can do that:

1.   RED	0.7.2.201610210841	org.robotframework.red	null
2.   Python version:
      ```
      (r2)misel@misel-amphinicy[~/projects/msbc-validation-tests]:
      [11:05:52] (master)* $ python --version
      Python 2.7.12
      ```
3.   Robot framework version:
      ```
      (r2)misel@misel-amphinicy[~/projects/msbc-validation-tests]:
      [11:05:56] (master)* $ pip list --format=columns | grep robot
      robotframework                          3.0    
      robotframework-databaselibrary          0.8.1  
      robotframework-extendedselenium2library 0.9.1  
      robotframework-ride                     1.5.2.1
      robotframework-selenium2library         1.8.0  
      robotframework-sshlibrary               2.1.2
      ```

**Comment by adam-wypych** [comment]
Could You show the file /home/misel/projects/msbc-validation-tests/vars/resource_vars.py ?

**Comment by mmesnjak** [comment]
Sure:
```
import provisioning_vars as _pv

mam_type = _pv.PI.get_type('ns0:MultipleAccessModulator')
vsats2mod_type = _pv.PI.get_type('ns0:VSatS2Modulator')
 
direction_type = _pv.PI.get_type('ns0:direction')
lcp_type = _pv.PI.get_type('ns0:LinkCalculationParameters')

LIST__RESOURCE_DATA = [
    mam_type(resourceName='Ma-Mod-1',
             supportedServiceClasses=['Satellite Internet Access - 3100 Forward'],
             resourceType='Modulator',
             resourceId=13,
             direction=direction_type('TRANSMIT'),
             remote=False,
             maxBookings=1,
             terminalId=0,
             maxTxBps=20000000,
             vsatSatNet='SatNet-1',
             vsatModem='HUB'
            ),
]
```

There is more objects in the list but RED acts the same in this case...

**Comment by adam-wypych** [comment]
Thanks now I see what is the issue :)

**Comment by mmesnjak** [comment]
Great!
Thanks for the agile support.

**Comment by adam-wypych** [comment]
This is just an description of problem.

The issue is connected to nature of WSDL/XML itself and the special handling regarding dynamic XML object creation declared in https://github.com/mvantellingen/python-zeep/blob/master/src/zeep/xsd/valueobjects.py class `CompoundValue`.

When I created script `InternalVarClass.py` like below:
```python
from collections import OrderedDict

''' below copied from http://stackoverflow.com/questions/15247075/how-can-i-dynamically-create-derived-classes-from-a-base-class/15247892#15247892 '''
class BaseClass(object):
	def __init__(self, classtype):
		self._type = classtype

def ClassFactory(name, argnames, BaseClass=BaseClass):
	def __init__(self, **kwargs):
		for key, value in kwargs.items():
			# here, the argnames variable is the one passed to the
			# ClassFactory call
			if key not in argnames:
				raise TypeError("Argument %s not valid for %s" % (key, self.__class__.__name__))
			setattr(self, key, value)
		BaseClass.__init__(self, name[:-len("Class")])
	newclass = type(name, (BaseClass,),{"__init__": __init__})
	return newclass

''' copied from https://github.com/mvantellingen/python-zeep/blob/master/src/zeep/xsd/valueobjects.py '''
class CompoundValue(object):

	def __init__(self, *args, **kwargs):
		values = OrderedDict()
		self.__values__ = values

	def __contains__(self, key):
		return self.__values__.__contains__(key)

	def __len__(self):
		return self.__values__.__len__()

	def __iter__(self):
		return self.__values__.__iter__()

	def __repr__(self):
		return str(self.__values__)

	def __delitem__(self, key):
		return self.__values__.__delitem__(key)

	def __getitem__(self, key):
		return self.__values__[key]

	def __setitem__(self, key, value):
		self.__values__[key] = value

	def __setattr__(self, key, value):
		if key.startswith('__') or key in ('_xsd_type', '_xsd_elm'):
			return super(CompoundValue, self).__setattr__(key, value)
		self.__values__[key] = value

	def __getattribute__(self, key):
		if key.startswith('__') or key in ('_xsd_type', '_xsd_elm'):
			return super(CompoundValue, self).__getattribute__(key)
		try:
			return self.__values__[key]
		except KeyError:
			raise AttributeError(
				"%s instance has no attribute '%s'" % (
					self.__class__.__name__, key))

	def __deepcopy__(self, memo):
		new = type(self)()
		new.__values__ = copy.deepcopy(self.__values__)
		for attr, value in self.__dict__.items():
			if attr != '__values__':
				setattr(new, attr, value)
		return new
_SpecialClass = ClassFactory("SpecialClass", "a b c".split())
X = type("X", (object,), dict(a=1))
LIST__v = [X(), _SpecialClass(a=10), CompoundValue()]
```

And after I use `${X}` or `${v}` variable in test suite file I get an error from XML-RPC server:
```robotframework
*** Settings ***
Variables	InternalVarClass.py

*** Test Cases ***
Test
	Log	${v}
	Log	${X}
```

From this it seems that `CompoundValue` is the problematic object additionally even normal object can't be send via current solution, so it will be solved like it is in Debug mode.

**Comment by adam-wypych** [comment]
The correction for this issue should be available in new upcoming release.

**Comment by mmesnjak** [comment]
Hi,

Just one quick question... Is there a roadmap for the next planned releases?

**Comment by KrzysztofJozefowicz** [comment]
Hi
We do not have strict road plan, as in agile it changes a lot. Due to holiday season (Christmas & New Year) we will not release any new version in 2016. We have plenty of small issues to fix which were postponed for some time (#57, content assist for resource/libs names, include/exclude validation for files, WITH NAME code coloring and assist update, error windows from libs resizable and more informative, quick fix for missing resources, comment multilines in tables) but some bigger features as code format (similar to PyDev rules) ,error suppression for selected item and make Robot&Python debug more user friendly.  Of course,this is not fixed list, there will be some shifting due to internal&external needs. 
Likely we will make next release around mid/end of January 2017.

If you have any ideas or needs, please provide github issue.
